File: typescript-kata/src/trees/inOrderTraversal.ts
import { BinaryNode } from "./node";

function walk(curr: BinaryNode<number> | undefined, path: number[]): number[] {
    if (!curr) {
        return path
    }

    // Pre
    //
    // Recursive
    walk(curr.left, path)
    path.push(curr.val)
    walk(curr.right, path)

    // Post
    return path
}

export default function InOrderTraversal(head: BinaryNode<number>): number[] {
    return walk(head, [])
}

File: go-kata/internal/trees/binary_tree/in_order_traversal.go
package binary_tree

func walkInOrder[T any](curr *BinaryNode[T], path *[]T) []T {
	if curr == nil {
		return *path
	}

	// Pre

	// Recursion
	walkInOrder(curr.Left, path)
	*path = append(*path, curr.Val)
	walkInOrder(curr.Right, path)

	// Post
	return *path
}

func InOrderTraversal[T any](head *BinaryNode[T]) []T {
	path := []T{}
	return walkInOrder(head, &path)
}
